core-backend/common/src/funcs.rs

```diff
@@ -0,0 +1,659 @@
+                delay,
+            )
+            .map_err(Into::into)
+    }
+
+    #[host(fallible, cost = RuntimeCosts::ReservationSendCommit)]
+    pub fn reservation_send_commit(
+        ctx: &mut R,
+        handle: u32,
+        rid_pid_value_ptr: u32,
+        delay: u32,
+    ) -> Result<(), R::Error> {
+        let read_rid_pid_value = ctx.register_read_as(rid_pid_value_ptr);
+        let TwoHashesWithValue {
+            hash1: reservation_id,
+            hash2: destination,
+            value,
+        } = ctx.read_as(read_rid_pid_value)?;
+
+        ctx.ext_mut()
+            .reservation_send_commit(
+                reservation_id.into(),
+                handle,
+                HandlePacket::new(destination.into(), Default::default(), value),
+                delay,
+            )
+            .map_err(Into::into)
+    }
+
+    #[host(fallible, cost = RuntimeCosts::Read, err_len = LengthBytes)]
+    pub fn read(ctx: &mut R, at: u32, len: u32, buffer_ptr: u32) -> Result<(), R::Error> {
+        let (buffer, mut gas_left) = ctx.ext_mut().read(at, len)?;
+        let buffer = buffer.to_vec();
+
+        let write_buffer = ctx.register_write(buffer_ptr, len);
+        ctx.memory_manager_write(write_buffer, &buffer, &mut gas_left)?;
+
+        ctx.ext_mut().set_gas_left(gas_left);
+        Ok(())
+    }
+
+    #[host(cost = RuntimeCosts::Size)]
+    pub fn size(ctx: &mut R, size_ptr: u32) -> Result<(), R::Error> {
+        let size = ctx.ext_mut().size()? as u32;
+
+        let write_size = ctx.register_write_as(size_ptr);
+        ctx.write_as(write_size, size.to_le_bytes())
+            .map_err(Into::into)
+    }
+
+    #[host(cost = RuntimeCosts::Exit)]
+    pub fn exit(ctx: &mut R, inheritor_id_ptr: u32) -> Result<(), R::Error> {
+        let read_inheritor_id = ctx.register_read_decoded(inheritor_id_ptr);
+        let inheritor_id = ctx.read_decoded(read_inheritor_id)?;
+        Err(ActorTerminationReason::Exit(inheritor_id).into())
+    }
+
+    #[host(fallible, cost = RuntimeCosts::StatusCode, err_len = LengthWithCode)]
+    pub fn status_code(ctx: &mut R) -> Result<(), R::Error> {
+        ctx.ext_mut().status_code().map_err(Into::into)
+    }
+
+    #[host(cost = RuntimeCosts::Alloc)]
+    pub fn alloc(ctx: &mut R, pages: u32) -> Result<u32, R::Error> {
+        let res = ctx.alloc(pages);
+        let res = ctx.process_alloc_func_result(res)?;
+
+        let page = match res {
+            Ok(page) => {
+                log::trace!("Alloc {pages:?} pages at {page:?}");
+                page.raw()
+            }
+            Err(err) => {
+                log::trace!("Alloc failed: {err}");
+                u32::MAX
+            }
+        };
+        Ok(page)
+    }
+
+    pub fn free(ctx: &mut R, page_no: u32) -> Result<i32, R::Error> {
+        syscall_trace!("free", page_no);
+
+        let page = WasmPage::new(page_no).map_err(|_| R::unreachable_error_code())?;
```

Comment:
I guess we must move it inside the `ctx.run_any`, because we receive an input, which fails validation. That's a common error situation. Otherwise we will have `termination_reason` be `Success` with `HostError`/`Trap` as a result of calling the wasm entry point

Reply:
The correct behaviour is return `1` from `free` in this case. So, we must move it inside closure and match it with `res`